crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -825,19 +827,64 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_for_statement(&mut self, for_statement: &ast::StmtFor) {
         let ast::StmtFor {
             range: _,
-            target,
+            target: _,
             iter,
             body,
             orelse,
             is_async: _,
         } = for_statement;
 
         self.infer_expression(iter);
-        self.infer_expression(target);
+        self.infer_definition(for_statement);
```

Comment:
This will need to become a lookup by a specific name node, rather than the overall for statement node, so that we can lookup the right definition for each symbol. For now we can just match on the target and lookup the definition by it if its a name node, and leave unpacking as a TODO.